Rewrite playlists to not inherit `RoomSubScreen`

smoogipoo opened the pull request:
This is a full rewrite of `PlaylistsRoomSubScreen` as a standalone screen that no longer inherits from `RoomSubScreen`. What I want to achieve primarily is the eventual separation of multiplayer and playlists structures/operation, so I plan to do the same with `MultiplayerMatchSubScreen` next.

It's a bit more than moving around code from the base class, but a bit less than totally redesigning every component. So while some things could be improved such as embedding the background box into `PlaylistsRoomFooter`, I haven't done it to keep the changes focused towards my immediate goal.

I suggest going through the screen with a fresh set of eyes.

smoogipoo committed:
Rewrite playlists to not inherit `RoomSubScreen`

smoogipoo committed:
Fix not properly selecting the first playlist item

smoogipoo committed:
Remove unnecessary update, document other usage

smoogipoo committed:
Rewrite/optimise layout

Reducing the amount of nesting to make things more readable.

smoogipoo committed:
Always update selection when anything changes

If style changes, then we also need to re-validate mods. Thus, we should
just update the entire selection anyway, and merge the "gameplay state"
into it for simplicity.

smoogipoo committed:
Merge branch 'master' into room-subscreen-split

smoogipoo committed:
Use margins to remove padding from hidden elements

If one of these elments is hidden, the following spacing element is
expected to be hidden too. Simplest is to use margins.

Old implementation already did this.

smoogipoo committed:
Fix inverted condition

bdach commented on osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomSubScreen.cs:
```diff
@@ -2,360 +2,804 @@
+            updateSelectionState();
+        }
+
+        public override bool OnExiting(ScreenExitEvent e)
+        {
+            if (!ensureExitConfirmed())
+                return true;
+
+            RoomManager?.PartRoom();
+
+            endHandlingTrack();
+            return base.OnExiting(e);
+        }
```
**Blocker**: `OnExiting()` should clear global mods, like `RoomSubScreen` does. Without doing that, you can have something like the following happen:

https://github.com/user-attachments/assets/46051de5-706a-42b1-8d14-913d317a1565

The issue is the track speeding up again after exiting the room subscreen (where it correctly resets to 1x speed) and then re-entering it to create a new room.

bdach commented:
Operationally, please advise how to proceed with order of merge of this and the series starting from https://github.com/ppy/osu/pull/31637, as I don't see an explicit dependency listed in the OP, but I also don't foresee a world where the two don't conflict even a little bit.

smoogipoo commented:
I worked on these changes separately, so I don't mind them being merged in any order. I'll have to resolve conflicts either way.

smoogipoo committed:
Merge branch 'master' into room-subscreen-split

smoogipoo committed:
Reset mods when exiting to lounge

To be hoenst, this is all quite abusive but I'm not sure how to do it in
any better way. I especially dislike the fact that the screen's
bindables are disabled yet the screen itself is allowed to set their
value because they're `LeasedBindable`s.

But I feel better doing this in a central location like this
`ScreenStack` than in the rooms... Otherwise this specific behaviour
would have to be replicated in the multiplayer screen when it goes
through the same refactoring.

smoogipoo committed:
Block input to screen during gameplay

smoogipoo commented on osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomSubScreen.cs:
```diff
@@ -2,360 +2,804 @@
+            beginHandlingTrack();
+        }
+
+        public override void OnSuspending(ScreenTransitionEvent e)
+        {
+            endHandlingTrack();
+            base.OnSuspending(e);
+        }
```
This is completely _fucked_.

Normally, input would be blocked by `ScreenStack` when a new screen is pushed: https://github.com/ppy/osu-framework/blob/e7f3343f3022dd32f2c8264e7cc421bf5414a4b0/osu.Framework/Screens/ScreenStack.cs#L375  
But, because this is inside a _child_ `ScreenStack`, this doesn't take effect fast enough.

This is entirely relying on the absolute WTF that is `OnlinePlayScreen` and in particular the following:
https://github.com/ppy/osu/blob/f6cf63edaea5f719b1ce177c8da0dae52e4d2647/osu.Game/Screens/OnlinePlay/OnlinePlayScreen.cs#L120-L148

What.

I've applied what I believe is a better fix via https://github.com/ppy/osu/pull/31882/commits/51f794a56675ee119b57d3b2f43cb012d7f0cf2a. Let me know if you agree.

smoogipoo replied:
Yeah, you're not wrong. I wouldn't have done it had I not had multiplayer failures down this same path - it opens mod select, then runs another test that tries to open mod select again, and fails because one's already in the overlay and async disposal hasn't completed yet.

May end up writing a test for this, if it's somewhat easy to do...

smoogipoo commented on osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomSubScreen.cs:
```diff
@@ -2,360 +2,804 @@
+                userRuleset.Value = null;
+            }
+
+            // Reset beatmap style when no longer from the same beatmap set.
+            if (userBeatmap.Value != null && userBeatmap.Value.BeatmapSet!.OnlineID != item.Beatmap.BeatmapSet!.OnlineID)
+                userBeatmap.Value = null;
+
+            IBeatmapInfo gameplayBeatmap = userBeatmap.Value ?? item.Beatmap;
+
+            // Reset ruleset style when no longer valid for the beatmap.
+            if (userRuleset.Value != null)
+            {
+                int beatmapRuleset = gameplayBeatmap.Ruleset.OnlineID;
+                if (beatmapRuleset > 0 && userRuleset.Value.OnlineID != beatmapRuleset)
+                    userRuleset.Value = null;
+            }
```
Looking into this one actually exposed a few larger issues. For example that referencing `item.Beatmap.BeatmapSet!.OnlineId` is wrong because there is no beatmap set...

Of course, this immediately blocks the PR. I'm trying to see if any of this can still be preserved, but the easiest path may be to do the same on `master` and indiscriminately reset user style when changing items.

smoogipoo replied:
I decided to take a bit of a step back on this one and act on my above comment to simplify the process a bit. I've documented it as part of https://github.com/ppy/osu/pull/31882/commits/0bdbaa872ceb6c8d18c68aaf175dcc95548e3a87, please have a read and see if it makes sense.

tl;dr:
- When a new playlist item is selected, beatmap style is always reset.
- Ruleset style is preserved only if it can be taken on for the new beatmap.
- Mods are preserved as much as possible, although right now this is based on `Mod` types rather than converting through `APIMod`/trying to create corresponding mods for other rulesets.

bdach replied:
I've read that commit like 3 times and it seems fine but your comments appear to imply that the previous code was in some way terminally broken and I don't particularly understand why. I'm ok just doing empirical testing and trusting that you've been sitting in all this much longer than I and probably have a better view on it though.

smoogipoo committed:
Add forgotten partials

smoogipoo committed:
Simplify user style validation

This commit takes half a step back.

Rather than trying to preserve the beatmap style across selections, it
probably makes more sense to always reset it. It is more likely for
playlists to be dominated by unique beatmap sets anyway. Doing this
already heavily simplifies the process.

Given the above, we can rely on the fact that the user beatmap and
ruleset are selected together, and are thus implicitly validated as of
exiting the style selection screen.

It follows that the only extra validation we need to do is to make sure
that the user's ruleset is valid for the playlist item.

This even makes the mod structure, which is a bit of an unfortunate
scenario, somewhat tenable to look at. The user mods only need to be
validated when either the playlist item or the user ruleset changes. If
we ever move the ruleset selection into the style selection screen, then
we can also remove the latter of the validations with similar reason as
the ruleset.

smoogipoo committed:
Add a slew of tests

smoogipoo committed:
Hide mod select overlay on exit

Only discovered this when running multiplayer tests on the upcoming
rewrite for that screen.

smoogipoo committed:
Also hide when suspending

bdach commented:
Is this still blocked?

smoogipoo commented:
Nah, can be reviewed again.

bdach commented on osu.Game/Screens/OnlinePlay/OnlinePlaySubScreenStack.cs:
```diff
@@ -12,16 +13,31 @@ protected override void ScreenChanged(IScreen prev, IScreen? next)
 
-            bool disallowChanges = osuScreen.DisallowExternalBeatmapRulesetChanges;
+            // First, emulate the behaviour of DisallowExternalBeatmapRulesetChanges to disable toolbar buttons.
+            osuNext.Beatmap.Disabled = osuNext.DisallowExternalBeatmapRulesetChanges;
```
Test failure here, unsure how relevant as this isn't really new code

smoogipoo replied:
Fixed in https://github.com/ppy/osu/pull/31882/commits/dbf571950c3f8b105c17fdcad5495a5f4a4738a1

Unlikely to be hit in practice, though not impossible. Can be reproed in test with:

```diff
diff --git a/osu.Game.Tests/OnlinePlay/TestSceneOnlinePlaySubScreenStack.cs b/osu.Game.Tests/OnlinePlay/TestSceneOnlinePlaySubScreenStack.cs
index d463610034..3e98810052 100644
--- a/osu.Game.Tests/OnlinePlay/TestSceneOnlinePlaySubScreenStack.cs
+++ b/osu.Game.Tests/OnlinePlay/TestSceneOnlinePlaySubScreenStack.cs
@@ -1,7 +1,9 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System.Threading;
 using NUnit.Framework;
+using osu.Framework.Allocation;
 using osu.Framework.Graphics;
 using osu.Framework.Screens;
 using osu.Framework.Testing;
@@ -64,6 +66,12 @@ public ScreenWithExternalBindableDisablement(bool disableBindables)
             {
                 DisallowExternalBeatmapRulesetChanges = disableBindables;
             }
+
+            protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnlyDependencyContainer parent)
+            {
+                Thread.Sleep(1000);
+                return base.CreateChildDependencies(parent);
+            }
         }
 
         private partial class ScreenWithMod : OsuScreen

```

smoogipoo committed:
Add failing test

smoogipoo committed:
Fix ruleset style not reset when freestyle disabled

smoogipoo committed:
Fix typo

smoogipoo committed:
Safguard setting states to not depend on screen load timing

smoogipoo commented on osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomSubScreen.cs:
```diff
@@ -2,362 +2,849 @@
             {
-                if (this.IsCurrentScreen())
-                    this.Push(new PlaylistsSongSelect(Room));
-            },
-        };
+                // If freestyle is active, attempt to preserve the user ruleset style but only if the online item is from the osu! ruleset
+                // (i.e. the beatmap is generally always convertible to the current ruleset, excluding custom rulesets).
+                if (item.RulesetID > 0)
+                    UserRuleset.Value = null;
+            }
```
Good catch. Fixed in https://github.com/ppy/osu/pull/31882/commits/eb15217a3661ec09bd0aec837450d6fe45467548, test added in https://github.com/ppy/osu/pull/31882/commits/d220c198858540554724f4fa4408966a3aaea6ce.

Server-side validation is probably a task @ppy/team-web can do better than me (if freestyle is disabled, user ruleset/beatmap must match playlist item).

nanaya replied:
ruleset check has been there for a long time unless you mean you tested that it doesn't do that...

nanaya replied:
ruleset_id parameter is ignored for non freestyle at the moment and my guess is the ruleset_id doesn't match what's being actually played

bdach replied:
> ruleset_id parameter is ignored for non freestyle at the moment

That sounds like the fail case here yeah.

bdach commented on osu.Game/Screens/OnlinePlay/OnlinePlaySubScreenStack.cs:
```diff
@@ -1,27 +1,74 @@
+
+        protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnlyDependencyContainer parent)
+        {
+            // Bindables are leased by the OnlinePlayScreen, but pulled locally in order ot not rely on screen load timings.
```
```suggestion
            // Bindables are leased by the OnlinePlayScreen, but pulled locally in order to not rely on screen load timings.
```

bdach commented on osu.Game/Screens/OnlinePlay/OnlinePlaySubScreenStack.cs:
```diff
@@ -1,27 +1,74 @@
     {
+        private OsuScreenDependencies dependencies = null!;
+
+        // Note - these are required to be unbound on disposal.
```
I dunno if it's just me but I found this comment a bit cryptic as well, so maybe better to be a little more explicit (if I'm even understanding the intent here correctly?)

```suggestion
        // Note - these bindables must be stored to fields of this component to be correctly unbound on disposal.
```

bdach reviewed (dismissed):
Think I've done enough testing on this, if there is something broken still (which I don't rule out...) I guess we'll just dogfood / fix in post

smoogipoo committed:
Apply commenting adjustments

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy commented:
something seems quite wrong with the background handling:

https://github.com/user-attachments/assets/ff6cc222-374f-46ce-9861-d099b9ad1679

peppy reviewed (changes requested):
as commented

smoogipoo committed:
Fix missing base dependencies

smoogipoo commented:
Re: backgrounds. The same issue is present in master:
```diff
diff --git a/osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs b/osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs
index ef7c1747e9..a88697d7af 100644
--- a/osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs
+++ b/osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs
@@ -21,6 +21,8 @@ public abstract partial class OnlinePlayBackgroundScreen : BackgroundScreen
         [BackgroundDependencyLoader]
         private void load()
         {
+            Thread.Sleep(2000);
+
             switchBackground(new PlaylistItemBackground(playlistItem));
         }
 

```
It's just that `master` was able to hide it better due to pushing without animation (fixed in https://github.com/ppy/osu/pull/31882/commits/148fe5ca16e92f6ea783f71cffc12ffe86135486). Besides that, this PR generally doesn't touch backgrounds.

The whole background thing needs a reconsideration of its own, because there's multiple levels of async through ScreenStacks here. It's not an easy fix.

peppy committed:
Fix loading layer not showing when closing room

peppy commented on osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomSubScreen.cs:
```diff
@@ -2,362 +2,851 @@
+        /// </summary>
+        private void onSelectedItemChanged(ValueChangedEvent<PlaylistItem?> e)
+        {
+            if (e.NewValue is not PlaylistItem item)
```
Not a huge fan of using `is not` in place of null checks. My brain short circuits when I read this. To clarify, I'm fine if it's being used to do an actual cast.

```diff
diff --git a/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomSubScreen.cs b/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomSubScreen.cs
index f8dd9cd3d9..bd0024a652 100644
--- a/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomSubScreen.cs
+++ b/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomSubScreen.cs
@@ -526,19 +526,19 @@ private void updatePollingRate()
         /// <summary>
         /// Responds to changes in <see cref="SelectedItem"/> to validate the user style and update the global gameplay state.
         /// </summary>
-        private void onSelectedItemChanged(ValueChangedEvent<PlaylistItem?> e)
+        private void onSelectedItemChanged(ValueChangedEvent<PlaylistItem?> item)
         {
-            if (e.NewValue is not PlaylistItem item)
+            if (item.NewValue == null)
                 return;
 
             // Always resetting the user beatmap style when a new item is selected is most intuitive.
             UserBeatmap.Value = null;
 
-            if (item.Freestyle)
+            if (item.NewValue.Freestyle)
             {
                 // If freestyle is active, attempt to preserve the user ruleset style but only if the online item is from the osu! ruleset
                 // (i.e. the beatmap is generally always convertible to the current ruleset, excluding custom rulesets).
-                if (item.RulesetID > 0)
+                if (item.NewValue.RulesetID > 0)
                     UserRuleset.Value = null;
             }
             else
@@ -554,16 +554,16 @@ private void onSelectedItemChanged(ValueChangedEvent<PlaylistItem?> e)
         /// </summary>
         private Mod[] listAllowedMods()
         {
-            if (SelectedItem.Value is not PlaylistItem item)
+            if (SelectedItem.Value == null)
                 return [];
 
-            RulesetInfo gameplayRuleset = UserRuleset.Value ?? rulesets.GetRuleset(item.RulesetID)!;
+            RulesetInfo gameplayRuleset = UserRuleset.Value ?? rulesets.GetRuleset(SelectedItem.Value.RulesetID)!;
             Ruleset rulesetInstance = gameplayRuleset.CreateInstance();
 
-            if (item.Freestyle)
+            if (SelectedItem.Value.Freestyle)
                 return rulesetInstance.AllMods.OfType<Mod>().Where(m => ModUtils.IsValidFreeModForMatchType(m, room.Type)).ToArray();
 
-            return item.AllowedMods.Select(m => m.ToMod(rulesetInstance)).ToArray();
+            return SelectedItem.Value.AllowedMods.Select(m => m.ToMod(rulesetInstance)).ToArray();
         }
 
         /// <summary>
@@ -580,9 +580,11 @@ private void validateUserMods()
         /// </summary>
         private void updateGameplayState()
         {
-            if (!this.IsCurrentScreen() || SelectedItem.Value is not PlaylistItem item)
+            if (!this.IsCurrentScreen() || SelectedItem.Value == null)
                 return;
 
+            PlaylistItem item = SelectedItem.Value;
+
             IBeatmapInfo gameplayBeatmap = UserBeatmap.Value ?? item.Beatmap;
             RulesetInfo gameplayRuleset = UserRuleset.Value ?? rulesets.GetRuleset(item.RulesetID)!;
             Ruleset rulesetInstance = gameplayRuleset.CreateInstance();
@@ -638,9 +640,11 @@ private void updateGameplayState()
         /// </summary>
         private void startPlay()
         {
-            if (!this.IsCurrentScreen() || SelectedItem.Value is not PlaylistItem item)
+            if (!this.IsCurrentScreen() || SelectedItem.Value == null)
                 return;
 
+            PlaylistItem item = SelectedItem.Value;
+
             // Required for validation inside the player.
             RulesetInfo gameplayRuleset = UserRuleset.Value ?? rulesets.GetRuleset(item.RulesetID)!;
             IBeatmapInfo gameplayBeatmap = UserBeatmap.Value ?? item.Beatmap;
@@ -661,7 +665,7 @@ private void startPlay()
         /// </summary>
         private void showUserModSelect()
         {
-            if (!this.IsCurrentScreen() || SelectedItem.Value is not PlaylistItem)
+            if (!this.IsCurrentScreen() || SelectedItem.Value == null)
                 return;
 
             userModsSelectOverlay.Show();
@@ -672,10 +676,10 @@ private void showUserModSelect()
         /// </summary>
         private void showUserStyleSelect()
         {
-            if (!this.IsCurrentScreen() || SelectedItem.Value is not PlaylistItem item)
+            if (!this.IsCurrentScreen() || SelectedItem.Value == null)
                 return;
 
-            this.Push(new PlaylistsRoomFreestyleSelect(room, item)
+            this.Push(new PlaylistsRoomFreestyleSelect(room, SelectedItem.Value)
             {
                 Beatmap = { BindTarget = UserBeatmap },
                 Ruleset = { BindTarget = UserRuleset }

```

smoogipoo replied:
It's going to take me a while to get used to this, but it's small enough that I'm going to apply it here. These days I'm finding it cleaner to write `if (aNullable is A a) { doSomething(a); }` than `doSomething aNullable.Value`, and this is probably a direct consequence of that.

My only request is that this is applied consistently, and it wasn't in `listAllowedMods()` in the above diff.

peppy replied:
Yeah the value part is annoying, sure. Going to prefer the local for now rather than `is` though.

bdach replied:
For what it's worth I find the "`is` is also a null-check" syntax not great as well because I mentally associate it with a typecast and thus it looks redundant if you don't think about it for like 5 seconds every time.

Also you only have to type `.Value` if the thing being null-checked-via-is is also a value type, which is a constant weird duality in C#'s typing.

smoogipoo committed:
Don't use `is` for null-checks

smoogipoo commented on osu.Game/Screens/OnlinePlay/Playlists/PlaylistsRoomSubScreen.cs:
```diff
@@ -2,362 +2,857 @@
+        /// <summary>
+        /// Validates the user mod style against the selected item and ruleset style.
+        /// </summary>
+        private void validateUserMods()
```
I split this out on purpose in https://github.com/ppy/osu/pull/31882/commits/0bdbaa872ceb6c8d18c68aaf175dcc95548e3a87, and I disagree that combining it simplifies anything as a reader.

The only validation that should be performed, ideally, is when the selected item is changed. User mods is just a special case because we use an overlay in the screen detached from the ruleset/beatmap "style" select.

This screen is linear in operation:
- Selected item changed -> attempt to transfer user style except beatmap across -> update global state
- Beatmap availability changed -> may now be able to get a WorkingBeatmap -> update global state
- User beatmap changed -> update global state
- User ruleset changed -> consider special case from the above -> update global state
- User mods changed -> update global state

As I see it, combining the two methods increases complexity as there's recursion to keep in mind whereby "setting the new global state" means also validating the user mod style, which also sets a new global state itself. All that, on top of the _existing_ validation when you select a new playlist item.

peppy replied:
Just going to look away from this for now then. I guess I misunderstood what the method does since validation methods are usually noops and can be called as often as required.

> User mods is just a special case because we use an overlay in the screen detached from the ruleset/beatmap "style" select.

Can I ask for an inline explanation of this though? Since I didn't realise this while reading the code and it seems like a good detail to be explained.

smoogipoo replied:
Added in https://github.com/ppy/osu/pull/31882/commits/9c0f3f9bef31fb3e541eb37c4b0fc0e290bf55c3

smoogipoo committed:
Describe special case of user mod validation
